### PR TITLE
remove metadata from UCM

### LINKED ISSUE #1060: discuss swapping out Star for Relation4 in Branch0.terms/types
`Relation4 r n t v` is a fully indexed version of `Set (r, n, t, v)`, whereas `Star3 r n t (t,v)` is a bidirectionally indexed version of `(Set (r,n), Set (r, t), Set (r, (t, v)))`.

This would allow us to associate metadata with individual names rather than only at the enclosing namespace: e.g:
> _#r_ has name _n_ with metadata (t,v)
> _#r has name _n2_ with metadata (t,v2)

The `Star3` version denotes:
> _#r_ has name _n_
> _#r_ has name _n2_
> _#r_ has metadata of type t
> _#r_ has metadata (t, v)
> _#r_ has metadata (t, v2)

This doesn't have to affect the serialization format initially; we can convert during serialization; but we wouldn't have to use both within `Branch.hs`.

P.S. This is independent of the "one-off" branch concern. In the alternate one-off formulation, a `n` is implied and only `Relation3 r t v` is needed.

### LINKED ISSUE #1062: BranchUtil.get{Term,Type}MetadataAt ignore the input NameSegment
I think we ultimately want to be able to associate different metadata to different names (#1060), but in the mean time, we suppose we don't want to return metadata when requesting a name that doesn't exist. 

They are currently called when handling `AliasTermI`, `AliasTypeI`, `doSlurpUpdates`, and `getLinks`.  We should think through each case and see if the caller can be simplified as well.

It could be that `get{Term,Type}MetadataAt` should take a `Maybe r`, to avoid an having to do a filter + fold in some cases?

### On the other hand

The functions _do_ take a Ref, so it may be that we don't ever construct such invalid queries.

### LINKED ISSUE #1128: base library metadata updates
- [ ] Update base library to have metadata types for authorship and licensing (an `Author` type and `License` type)
  - [ ] `.licenses.{MIT,BSD3,BSD2,Apache2...}`
  - [ ] `.authors.{pchiusano,runarorama,aryairani}`, and whoever else contributes to base
- [ ] Assign authorship and license to all definitions in base

### LINKED ISSUE #1398: autocomplete for `link` feels broken
image

What. I don't want to browse, I just want to autocomplete the word `metadata`. 😕 

Propose just doing forward completion of a suffixified namespace.  (Ref #1377)

### LINKED ISSUE #1455: metadata info in diff shouldn't blend into preceding type signature
```
.prs.runarorama> diff.namespace base merged

  Added definitions:
  
    1. Set.map           : (a ->{e} b) -> Set a ->{e} Set b (+3 metadata)
    2. Set.map.doc       : Doc (+2 metadata)
    3. Set.map.test      : [Result] (+3 metadata)
    4. Set.singleton     : a -> Set a (+3 metadata)
    5. Set.singleton.doc : Doc (+2 metadata)
  
    6. patch patch

.prs.runarorama> 
```
Let's put the `(+x metadata)` into their own column.

### LINKED ISSUE #1486: Default metadata generates too much history.
If you add a bunch of definitions, with default `Author` and `License` metadata, you get 3 entries in the history:

1. All the terms/types you're adding.
2. All the author links.
3. All the license links.

Would be great if only one history entry were generated on `add`.

### LINKED ISSUE #1492: unclear how to link an author on first PR
I've been playing around in Unison, and my experience has been that ramping up on the language has been smooth, but there are barriers to learning the daily workflows. I know that this is very much a work in progress, so I'll try to create helpful issues/PRs as I understand where I'm having trouble.

One issue I'm running into is attaching author metadata to my first `base` PR. The docs suggested the `create.author` command, which worked great. But _using_ the metadata isn't so clear to me.

The [Day-to-day development: creating and merging pull requests](https://www.unisonweb.org/docs/codebase-organization/#day-to-day-development-creating-and-merging-pull-requests) docs don't mention attaching author metadata. But I had seen the [Configuration](https://www.unisonweb.org/docs/configuration) section, so I suspected that this was something that I should do. When I followed these instructions I ran into a couple of questions/issues:

1. I was in a `.prs._myPR` namespace when I ran `create.author` command. Should I do this separately for each project that I work on, or should I have an `author` object somewhere in a common place that I reuse across projects? The docs mention that I can `alias` or `copy` the author object, but I don't know what that would look like.
2. When I follow the instructions on the configuration page and run `add` I get an error message `I couldn't find any metadata matching ._base.trunk.metadata.authors.ceedubs`. I suspect that this is because my `author` metadata isn't actually in `trunk` yet - it only lives in my `.prs._myPR` namespace. Should I be referencing my `.prs._myPR` namespace in my config? Or should I just try to manually link the author object until I exist in `trunk`? When I tried manually linking I had trouble figuring out how to do it.

### LINKED ISSUE #1494: docs and links commands don't match on suffixes
Some `ucm` commands like `view` will return results on matching suffixes (I think that it's more general than this but don't know what the constraints are).

However, the `docs` and `links` commands do not behave the same way. See example below.

```
.prs._genEither> view pairOf

  test.pairOf : '{Gen} a -> '{Gen} b -> '{Gen} (a, b)
  test.pairOf a b = '(!a, !b)

.prs._genEither> docs pairOf

  😶
  
  No results. Try using the `link` command to add metadata to a definition.

.prs._genEither> links pairOf

  😶
  
  No results. Try using the `link` command to add metadata to a definition.

.prs._genEither> docs test.pairOf

  Generates a pair given a generator for each of the elements.

.prs._genEither> links test.pairOf

  1. metadata.authors.runarorama           : Author
  2. metadata.licenses.unisoncomputing2020 : License
  3. test.pairOf.doc                       : Doc
  
  Tip: Try using `display 1` to display the first result or `view 1` to view its source.
```

### LINKED ISSUE #1803: ensure that `metadata.{IsTest, isTest, IsPropagated, isPropagated}` is added to codebase as a decl before user tests are added.
tldr:

```ucm
.> builtins.merge

  Done.

```
```unison
test> pass = [Ok "Passed"]
```

```ucm

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      pass : [Result]
  
  Now evaluating any watch expressions (lines starting with
  `>`)... Ctrl+C cancels.

    1 | test> pass = [Ok "Passed"]
    
    ✅ Passed Passed

```
```ucm
.> add

  ⍟ I've added these definitions:
  
    pass : [Result]

.> links pass

  1. #krhcbbspn6 : ❓ (missing a type for this definition)
  
  Tip: Try using `display 1` to display the first result or
       `view 1` to view its source.

.> display 1

  ⚠️
  
  The following names were not found in the codebase. Check your spelling.
    #krhcbbspn6

.>
```

We should never see "❓ (missing a type for this definition)", but we have managed to link metadata that isn't known to the codebase.  `#krhcbbspn6` is aka `isTest`; we should also have the type `IsTest` and the constructor #0 for it.

### LINKED ISSUE #2173: Codebase Server API: Include metadata in `/getDefinition`
Extend the `/getDefinition` codebase server API with metadata like author, license and data.

Completing this ticket will support this UI ticket: https://github.com/unisonweb/ui-core/issues/24

## Suggested response

To the already existing `/getDefinition` API, add a `metadata` field with the following shape:

```elm
type Metadata {
  authors: List Author,
  copyrightHolders: List CopyrightHolder,
  licenses: List License
  createdAt: String, -- ISO 8601 formatted date time 
  updatedAt: String -- ISO 8601 formatted date time
}
```

❗This is dependent on https://github.com/unisonweb/unison/issues/2205

### LINKED ISSUE #2185: Codebase Server API: `/namespaceMetadata`
A new endpoint for getting metadata information about a namespace (project). It is a unique aggregate of all metadata for sub namespaces and definitions.

This will support https://github.com/unisonweb/codebase-ui/issues/144 for rendering contributors and licenses across a project.

## Suggested Response Schema

```elm
type NamespaceMetadataResponse = { 
  authors: List Author,
  licenses: List Licenses,
  copyrightHolders: List CopyrightHolder
  createdAt: String, -- ISO 8601 formatted date time 
  updatedAt: String -- ISO 8601 formatted date time
}
```

### Alternative schema
```elm
type Metadata 
  = AuthorMetadata Author 
  | CopyrightHolderMetadata CopyrightHolder 
  | LicenseMetadata License

type NamespaceMetadataResponse = List Metadata
```

Alternatively this could be a `/metadata` endpoint that took in a list of hashes of metadata definitions to render—not sure how a Date works in that world?

this was factored out of https://github.com/unisonweb/unison/issues/2172, but could easily be merged back into it.

❗This is dependent on https://github.com/unisonweb/unison/issues/2205

### LINKED ISSUE #2205: Codebase Server: Support returning metadata in JSON
We want to be able to show metadata that exists in a Unison codebase in the UI.
This data is formatted as Unison code and we don't currently have a way, outside of SyntaxText to represent it.

Metadata include things like authors and licenses, but also dates about when a namespace/definition was last updated etc.

Metadata isn't a strict thing, so it can actually include all kinds of Unison types and terms. This effort explore metadata as a more formal construct.

A suggested schema (that can be used for several different endpoints that include metadata):

```elm
type Metadata {
  authors: List Author,
  copyrightHolders: List CopyrightHolder,
  licenses: List License
  createdAt: String, -- ISO 8601 formatted date time 
  updatedAt: String -- ISO 8601 formatted date time
}

type Author = {
  guid: String,
  username: String,
  name: String,
}

type CopyrightHolder = {
  guid: String,
  username: String,
  name: String,
}

type LicenseType = {
  name: String, -- "MIT"
  description: String
}

type License = {
  name: String,
  year: Nat,
  type: LicenseType
}
```

Required to support:
* https://github.com/unisonweb/unison/issues/2172
* https://github.com/unisonweb/unison/issues/2186
* https://github.com/unisonweb/unison/issues/2185

### LINKED ISSUE #2234: Codebase server backend should look for docs via linked metadata
Loose end from #2116 

At the moment, the way the endpoint finds docs is just via the naming convention that foo.doc is the docs for and anything called foo. It should probably check for linked doc metadata also, but that's not done currently.

### LINKED ISSUE #2314: can't override default metadata for deeper namespaces, only union
With this `~/.unisonConfig`, 
```
DefaultMetadata = [ ".config.author.arya"
                  , ".config.license.arya_mit_2021" ]

DefaultMetadata {
  cantstop = [".config.license.arya_mit_2021"]
  dogfood = ["#1lmvrom87b","#nr47uo1650"]
}
```

if I add something to `.dogfood`, I get the top-level default metadata as well as the `.dogfood` metadata.  I could imagine wanting that, but I didn't in this case; I wanted to apply only a unisoncomputing license (`#1lmvrom87b`) and not the `arya_mit_2021` license.

### LINKED ISSUE #2315: each default metadata is added in a new causal cons
e.g. in #2314, when I `add` my definition and got three default metadata, there were 4 causals total.  One for the add, and one for each of three default metadata in some arbitrary order.

### LINKED ISSUE #2417: Signed metadata
Would be nice to allow metadata to be cryptographically signed, so that, for instance, you could verify that someone with Alice’s private key associated some metadata with a hash.

### LINKED ISSUE #2451: Add default metadata according to term's namespace instead of UCM CWD
Currently the default metadata is chosen based on the namespace of your UCM session, not the location of the term being added. E.g. if I add the following:

```
.base.List.map = ...
```

And my ucm session is at `.>` It will use the metadata from `.>` instead of `.base.List` if default metadata is specified there.

### LINKED ISSUE #2648: Metadata.Star has some issues
1. It allows metadata on things that aren't named in the namespace — this is mitigated by not being possible to represent in the v2 codebase, but these invalid branches are presumably still being generated in memory
2. It doesn't allow different sets of metadata to be associated with different names in the same namespace.  Instead, it defines a set of metadata for a given reference/referent within a given namespace, independent of names.

It's not totally clear to me whether these are features or bugs, but according to parser-typechecker they are features and according to codebase-sqlite, they are bugs.

I don't think this merits any immediate action, because `Unison.Codebase.Branch` might undergo a more significant transformation anyway before we get to this, but I wanted to document this somewhere.

### LINKED ISSUE #3535: surprising interactions between metadata links and self-contained namespaces?
If you fork / move / alias a definition that has metadata on it into some namespace that doesn't have a name for that metadata, you can unwittingly create an external name dependency.

via discussion with @ceedubs

### PR DESCRIPTION
**Pull request by aryairani**
- [ ] remove `link` command
- [ ] add a new hash version that doesn't include metadata
- [ ] do a rehash migration to discard metadata
- [ ] remove metadata from `Branch0` 
- [ ] either do a rehash migration to remove old metadata, or print a warning when entering a namespace that contained metadata, saying that some older version of UCM would be needed to view it.
- [ ] clean up transcripts
- [ ] remove `unlink` command
- [ ] anything to do with author or license linking in ucm
- [ ] determine what to do re. author, license, and copyright info and document those (website + ucm as required)
- [x] revisit startup workflow to decide what we want to do about author, copyright holder, etc.
- [ ] bring user docs up to date with the above
- [x] #4416 
- [ ] #4448

Closes #3535, closes #2205, closes #2173, closes #2648, closes #2451, closes #2417, closes #2315, closes #2314, closes #2234, closes #2185, closes #1803, closes #1486, closes #1455, closes #1494, closes #1492, closes #1397, closes #1398, closes #1062, closes #1128, closes #1060

### CODE REVIEW

**Comment by mitchellwrosen** [comment]
Partially done in #4574, and partially tracked in #4582
